modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/MetadataComponentModelValidator.java

```diff
@@ -144,14 +144,15 @@ private void validateResolversName(ComponentModel model, MetadataResolverFactory
         });
   }
 
-  private void validateMetadataKeyId(ComponentModel model, MetadataResolverFactory resolverFactory,
+  private void validateMetadataKeyId(ExtensionModel extensionModel, ComponentModel model, MetadataResolverFactory resolverFactory,
                                      ProblemsReporter problemsReporter) {
     final String modelTypeName = capitalize(getComponentModelTypeName(model));
     Optional<MetadataKeyIdModelProperty> keyId = model.getModelProperty(MetadataKeyIdModelProperty.class);
     if (keyId.isPresent()) {
 
       if (resolverFactory.getOutputResolver() instanceof NullMetadataResolver
-          && !thereIsAnInputTypeResolverDefined(getAllInputResolvers(model, resolverFactory))) {
+          && !thereIsAnInputTypeResolverDefined(getAllInputResolvers(model, resolverFactory))
+          && isCompileTime(extensionModel)) {
```

Comment:
make this the very first condition so that mule startup doesn't go through unnecessary processing.